File: ios/LibVlcPlayerModule.swift
import AVKit
import ExpoModulesCore

private let playerEvents = [
    "onBuffering",
    "onPlaying",
    "onPaused",
    "onStopped",
    "onEncounteredError",
    "onDialogDisplay",
    "onTimeChanged",
    "onPositionChanged",
    "onESAdded",
    "onRecordChanged",
    "onSnapshotTaken",
    "onFirstPlay",
    "onForeground",
    "onBackground",
    "onPictureInPictureStart",
    "onPictureInPictureStop",
]

public class LibVlcPlayerModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoLibVlcPlayer")

        AsyncFunction("triggerNetworkAlert") {
            MediaPlayerManager.shared.localNetworkManager.triggerNetworkAlert()
        }

        Function("isPictureInPictureSupported") {
            AVPictureInPictureController.isPictureInPictureSupported()
        }

        OnDestroy {
            MediaPlayerManager.shared.onModuleDestroy()
        }

        OnAppEntersForeground {
            MediaPlayerManager.shared.keepAwakeManager.toggleKeepAwake()
            MediaPlayerManager.shared.onModuleForeground()
        }

        OnAppEntersBackground {
            MediaPlayerManager.shared.keepAwakeManager.deactivateKeepAwake()
            MediaPlayerManager.shared.onModuleBackground()
        }

        View(LibVlcPlayerView.self) {
            Events(playerEvents)

            Prop("source") { (view: LibVlcPlayerView, source: String?) in
                view.source = source
            }

            Prop("options", .init()) { (view: LibVlcPlayerView, options: [String]) in
                view.options = options
            }

            Prop("tracks") { (view: LibVlcPlayerView, tracks: Tracks?) in
                view.tracks = tracks
            }

            Prop("slaves", .init()) { (view: LibVlcPlayerView, slaves: [Slave]) in
                view.slaves = slaves
            }

            Prop("scale", MediaPlayerConstants.defaultPlayerScale) { (view: LibVlcPlayerView, scale: Float) in
                view.scale = scale
            }

            Prop("contentFit", .contain) { (view: LibVlcPlayerView, contentFit: VideoContentFit) in
                view.contentFit = contentFit
            }

            Prop("rate", MediaPlayerConstants.defaultPlayerRate) { (view: LibVlcPlayerView, rate: Float) in
                view.rate = rate
            }

            Prop("time", MediaPlayerConstants.defaultPlayerTime) { (view: LibVlcPlayerView, time: Int) in
                view.time = time
            }

            Prop("volume", MediaPlayerConstants.maxPlayerVolume) { (view: LibVlcPlayerView, volume: Int) in
                view.volume = volume
            }

            Prop("mute", false) { (view: LibVlcPlayerView, mute: Bool) in
                view.mute = mute
            }

            Prop("audioMixingMode", .auto) { (view: LibVlcPlayerView, audioMixingMode: AudioMixingMode) in
                view.audioMixingMode = audioMixingMode
            }

            Prop("repeat", false) { (view: LibVlcPlayerView, shouldRepeat: Bool) in
                view.shouldRepeat = shouldRepeat
            }

            Prop("autoplay", true) { (view: LibVlcPlayerView, autoplay: Bool) in
                view.autoplay = autoplay
            }

            Prop("pictureInPicture", false) { (view: LibVlcPlayerView, pictureInPicture: Bool) in
                view.pictureInPicture = pictureInPicture
            }

            OnViewDidUpdateProps { (view: LibVlcPlayerView) in
                view.initPlayer()
            }

            AsyncFunction("play") { (view: LibVlcPlayerView) in
                view.play()
            }

            AsyncFunction("pause") { (view: LibVlcPlayerView) in
                view.pause()
            }

            AsyncFunction("stop") { (view: LibVlcPlayerView) in
                view.stop()
            }

            AsyncFunction("seek") { (view: LibVlcPlayerView, value: Double, type: String?) in
                view.seek(value, type)
            }

            AsyncFunction("record") { (view: LibVlcPlayerView, path: String?) in
                view.record(path)
            }

            AsyncFunction("snapshot") { (view: LibVlcPlayerView, path: String) in
                view.snapshot(path)
            }

            AsyncFunction("postAction") { (view: LibVlcPlayerView, action: Int) in
                view.postAction(action)
            }

            AsyncFunction("dismiss") { (view: LibVlcPlayerView) in
                view.dismiss()
            }

            AsyncFunction("startPictureInPicture") { (view: LibVlcPlayerView) in
                try view.startPictureInPicture()
            }

            AsyncFunction("stopPictureInPicture") { (view: LibVlcPlayerView) in
                view.stopPictureInPicture()
            }
        }
    }
}

File: android/src/main/java/expo/modules/libvlcplayer/LibVlcPlayerModule.kt
package expo.modules.libvlcplayer

import android.app.Activity
import android.os.Build
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.libvlcplayer.constants.MediaPlayerConstants
import expo.modules.libvlcplayer.enums.AudioMixingMode
import expo.modules.libvlcplayer.enums.VideoContentFit
import expo.modules.libvlcplayer.managers.MediaPlayerManager
import expo.modules.libvlcplayer.records.Slave
import expo.modules.libvlcplayer.records.Tracks

private val PLAYER_EVENTS =
    arrayOf(
        "onBuffering",
        "onPlaying",
        "onPaused",
        "onStopped",
        "onEncounteredError",
        "onDialogDisplay",
        "onTimeChanged",
        "onPositionChanged",
        "onESAdded",
        "onRecordChanged",
        "onSnapshotTaken",
        "onFirstPlay",
        "onForeground",
        "onBackground",
        "onPictureInPictureStart",
        "onPictureInPictureStop",
    )

class LibVlcPlayerModule : Module() {
    private val activity: Activity
        get() = appContext.currentActivity ?: throw Exceptions.MissingActivity()

    override fun definition() =
        ModuleDefinition {
            Name("ExpoLibVlcPlayer")

            Function("isPictureInPictureSupported") {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
                    activity.packageManager.hasSystemFeature(
                        android.content.pm.PackageManager.FEATURE_PICTURE_IN_PICTURE,
                    )
            }

            OnCreate {
                MediaPlayerManager.onModuleCreate(appContext)
            }

            OnDestroy {
                MediaPlayerManager.onModuleDestroy()
            }

            OnActivityEntersForeground {
                MediaPlayerManager.keepAwakeManager.toggleKeepAwake()
                MediaPlayerManager.onModuleForeground()
            }

            OnActivityEntersBackground {
                MediaPlayerManager.keepAwakeManager.deactivateKeepAwake()
                MediaPlayerManager.onModuleBackground()
            }

            View(LibVlcPlayerView::class) {
                Events(PLAYER_EVENTS)

                Prop("source") { view: LibVlcPlayerView, source: String? ->
                    view.source = source
                }

                Prop("options", mutableListOf()) { view: LibVlcPlayerView, options: MutableList<String> ->
                    view.options = options
                }

                Prop("tracks") { view: LibVlcPlayerView, tracks: Tracks? ->
                    view.tracks = tracks
                }

                Prop("slaves", mutableListOf()) { view: LibVlcPlayerView, slaves: MutableList<Slave> ->
                    view.slaves = slaves
                }

                Prop("scale", MediaPlayerConstants.DEFAULT_PLAYER_SCALE) { view: LibVlcPlayerView, scale: Float ->
                    view.scale = scale
                }

                Prop("contentFit", VideoContentFit.CONTAIN) { view: LibVlcPlayerView, contentFit: VideoContentFit ->
                    view.contentFit = contentFit
                }

                Prop("rate", MediaPlayerConstants.DEFAULT_PLAYER_RATE) { view: LibVlcPlayerView, rate: Float ->
                    view.rate = rate
                }

                Prop("time", MediaPlayerConstants.DEFAULT_PLAYER_TIME) { view: LibVlcPlayerView, time: Int ->
                    view.time = time
                }

                Prop("volume", MediaPlayerConstants.MAX_PLAYER_VOLUME) { view: LibVlcPlayerView, volume: Int ->
                    view.volume = volume
                }

                Prop("mute", false) { view: LibVlcPlayerView, mute: Boolean ->
                    view.mute = mute
                }

                Prop("audioMixingMode", AudioMixingMode.AUTO) { view: LibVlcPlayerView, audioMixingMode: AudioMixingMode ->
                    view.audioMixingMode = audioMixingMode
                }

                Prop("repeat", false) { view: LibVlcPlayerView, repeat: Boolean ->
                    view.repeat = repeat
                }

                Prop("autoplay", true) { view: LibVlcPlayerView, autoplay: Boolean ->
                    view.autoplay = autoplay
                }

                Prop("pictureInPicture", false) { view: LibVlcPlayerView, pictureInPicture: Boolean ->
                    view.pictureInPicture = pictureInPicture
                }

                OnViewDidUpdateProps { view: LibVlcPlayerView ->
                    view.initPlayer()
                }

                OnViewDestroys { view: LibVlcPlayerView ->
                    MediaPlayerManager.unregisterExpoView(view)
                    view.destroyPlayer()
                }

                AsyncFunction("play") { view: LibVlcPlayerView ->
                    view.play()
                }

                AsyncFunction("pause") { view: LibVlcPlayerView ->
                    view.pause()
                }

                AsyncFunction("stop") { view: LibVlcPlayerView ->
                    view.stop()
                }

                AsyncFunction("seek") { view: LibVlcPlayerView, value: Double, type: String? ->
                    view.seek(value, type)
                }

                AsyncFunction("record") { view: LibVlcPlayerView, path: String? ->
                    view.record(path)
                }

                AsyncFunction("snapshot") { view: LibVlcPlayerView, path: String ->
                    view.snapshot(path)
                }

                AsyncFunction("postAction") { view: LibVlcPlayerView, action: Int ->
                    view.postAction(action)
                }

                AsyncFunction("dismiss") { view: LibVlcPlayerView ->
                    view.dismiss()
                }

                AsyncFunction("startPictureInPicture") { view: LibVlcPlayerView ->
                    view.startPictureInPicture()
                }
            }
        }
}
